packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -218,106 +224,270 @@ export class ModularChangeFamily
+		assert(context !== undefined, 0x8cc /* Should have context for every invalidated field */);
+		const { change1: fieldChange1, change2: fieldChange2, composedChange } = context;
+
+		const rebaser = getChangeHandler(this.fieldKinds, composedChange.fieldKind).rebaser;
+		const composeNodes = (child1: NodeId | undefined, child2: NodeId | undefined): NodeId => {
+			if (
+				child1 !== undefined &&
+				child2 !== undefined &&
+				getFromChangeAtomIdMap(crossFieldTable.newToBaseNodeId, child2) === undefined
+			) {
+				crossFieldTable.nodeIdPairs.push([child1, child2]);
+				setInChangeAtomIdMap(crossFieldTable.newToBaseNodeId, child2, child1);
+			}
+
+			return child1 ?? child2 ?? fail("Should not compose two undefined nodes");
+		};
+
+		const amendedChange = rebaser.compose(
+			fieldChange1,
+			fieldChange2,
+			composeNodes,
+			genId,
+			new ComposeManager(crossFieldTable, fieldChange, false),
+			revisionMetadata,
 		);
+		composedChange.change = brand(amendedChange);
+	}
+
+	private composeDiscoveredFields(
+		change1: ModularChangeset,
+		change2: ModularChangeset,
+		table: ComposeTable,
+		composedFields: FieldChangeMap,
+		composedNodes: ChangeAtomIdMap<NodeChangeset>,
+		composedNodeToParent: ChangeAtomIdMap<FieldId>,
+		nodeAliases: ChangeAtomIdMap<NodeId>,
+		genId: IdAllocator,
+		metadata: RevisionMetadataSource,
+	): void {
+		while (
+			table.nodeIdPairs.length > 0 ||
+			table.affectedBaseFields.length > 0 ||
+			table.affectedNewFields.length > 0
+		) {
+			// Note that the call to `composeNodesById` can add entries to `crossFieldTable.nodeIdPairs`.
+			for (const [id1, id2] of table.nodeIdPairs) {
+				this.composeNodesById(
+					change1.nodeChanges,
+					change2.nodeChanges,
+					composedNodes,
+					composedNodeToParent,
+					nodeAliases,
+					id1,
+					id2,
+					genId,
+					table,
+					metadata,
+				);
+			}
+
+			table.nodeIdPairs.length = 0;
+
+			this.composeAffectedFields(
+				table,
+				table.baseChange,
+				table.affectedBaseFields,
+				composedFields,
+				composedNodes,
+				true,
+				genId,
+				metadata,
+			);
+
+			this.composeAffectedFields(
+				table,
+				table.newChange,
+				table.affectedNewFields,
+				composedFields,
+				composedNodes,
+				false,
+				genId,
+				metadata,
+			);
+		}
+	}
+
+	private composeAffectedFields(
+		table: ComposeTable,
+		change: ModularChangeset,
+		affectedFields: BTree<FieldIdKey, boolean>,
+		composedFields: FieldChangeMap,
+		composedNodes: ChangeAtomIdMap<NodeChangeset>,
+		areBaseFields: boolean,
+		genId: IdAllocator,
+		metadata: RevisionMetadataSource,
+	): void {
+		for (const fieldIdKey of affectedFields.keys()) {
+			const fieldId = normalizeFieldId(fieldIdFromFieldIdKey(fieldIdKey), change.nodeAliases);
+			const fieldChange = fieldChangeFromId(change.fieldChanges, change.nodeChanges, fieldId);
+
+			if (
+				table.fieldToContext.has(fieldChange) ||
+				table.newFieldToBaseField.has(fieldChange)
+			) {
+				continue;
```

Comment:
What is the reason for skipping in this case?

Reply:
Yes, this function handles fields which were not part of the intersection of the two changesets but which need to be updated anyway. If we've already processed this field then either it is up to date or there is pending inval which will be handled in `processInvalidatedCompositions`.

Reply:
Your reply would make a great comment here.